### DISCUSSION TITLE
Bokeh in figures with glue

### DISCUSSION THREAD

**Discussion by bryanwweber** [executablebooks/meta#346] [Q&A]
Hi everyone! I'd like to be able to `glue` Bokeh graphs into a `figure` so that I can reference them in the text. Using the recommendation here: https://jupyterbook.org/interactive/interactive.html#bokeh of Bokeh's `output_notebook()` setup function, trying to do:

```python
p = figure()
# plot something...
glue("bokeh_figure", p)
```

and

````markdown
Some text.

```{glue:figure} bokeh_figure
:name: bokeh_figure

Caption
```
````

just shows the `repr` of the `figure()` instance, something like `Figure(id='1002', ...)`.

In looking at the implementation of `glue` in [`myst_nb.nb_glue`](https://github.com/executablebooks/MyST-NB/blob/6be37cd66d1b7668b16a4bab385222b58f38930d/myst_nb/nb_glue/__init__.py#L7) and the related IPython classes, it seems like the Bokeh figure might have to implement something like `__repr_svg__` or similar. At the moment, the mimebundle only includes `text/plain` and `text/html`. Is that correct, or is there something that Jupyter Book can do to be able to glue Bokeh plots into figures?

**Comment by choldgraf** [comment]
Hmmm, that is a good question, I think your intuition is right but perhaps @chrisjsewell has a better idea...you might be the first person that has tried this 🙂

**Comment by bryanwweber** [comment]
It is possible to do this by directly using the `IPython.display` function to put output into the cell metadata.

```python
import numpy as np
from bokeh.plotting import figure
from IPython.display import display as ipy_display
from bokeh.embed import components
import xml.etree.ElementTree as ET

np.random.seed(19680801)
N = 10
data = np.linspace(1, 10, 10)
p = figure()
p.circle(data, data)

script, div = components(p)

mime_prefix = "application/papermill.record/"
data = ET.fromstring(script)
htmlbundle = {mime_prefix + "text/html": div}
jsbundle = {mime_prefix + "application/javascript": data.text}

htmlmeta = {"scrapbook": dict(name="bokeh_html", mime_prefix=mime_prefix)}
jsmeta = {"scrapbook": dict(name="bokeh_js", mime_prefix=mime_prefix)}
ipy_display(htmlbundle, raw=True, metadata=htmlmeta)
ipy_display(jsbundle, raw=True, metadata=jsmeta)

# ```{glue:figure} bokeh_html
# :name: bokeh_figure
# 
# Caption
# ```

# ```{glue:} bokeh_js
# 
```

The hangup is adding the Javascript to load the Bokeh library to the HTML somewhere. I also want to look at the `export_file()` method from Bokeh to see if that includes the complete HTML document.

**Reply by bryanwweber** [reply]
So `export_file()` won't quite work, or at least, it's not any easier, because it provides a complete HTML document, including the `<html>` tag.

Including the Bokeh JS files can be done by putting them in the `_static` folder of the book. This includes it on every page, which isn't ideal, but better than not having it!

**Comment by bryanwweber** [comment]
Building on the previous approach, I added a `glue_bokeh()` function and `PasteBokeh` class to MyST-NB over here: https://github.com/bryanwweber/MyST-NB/pull/1 (the pull request is against my own fork for ease of file comparison; if it's worth pursuing I'm happy to make a PR against the upstream). Copying the PR comment from there:

This change adds the `glue_bokeh()` function to insert the Bokeh figure
HTML and JS components into the cell output. The `PasteBokeh` handler then
collects the appropriate glued output and returns both nodes to be
included simultaneously.

The Bokeh figure components are returned as strings from the components
method, so they are converted to IPython HTML and Javascript objects
before being passed to glue() so that they have the correct mimetypes
for the display formatter. In addition, creating the Javascript object
allows us to add the BokehJS library dependency which will be injected
into the head of the HTML file.

This approach seems to be fairly robust, insofar as the figure is pasted
properly. It does not yet support being pasted as a MyST-NB Figure
though. Disadvantages of this approach, as I see them, are:

1) It requires a separate function to glue Bokeh figures, which isn't a
   super great UI, if it can be avoided
2) If a page has multiple pasted Bokeh figures, it isn't clear whether
   the BokehJS library will be injected into the header multiple times,
   and I'm not familiar enough with HTML semantics to know whether
   loading the same script multiple times is actually a problem.

**Reply by bryanwweber** [reply]
Input looks like this:

![image](https://user-images.githubusercontent.com/4396228/120910367-59b85980-c64c-11eb-8dc6-c0683487e089.png)

and output looks like this:

![image](https://user-images.githubusercontent.com/4396228/120910372-689f0c00-c64c-11eb-8bb8-b6e24b2868dc.png)

**Reply by damianavila** [reply]
This is pretty interesting! Probably others with more experience in myst_nb can look at the PR and provide full feedback.
But just from a quick overview, I see you are importing Bokeh components and that add a new dependency to the project. So you can maybe import those conditionally and fails graciously if Bokeh is not installed in your environment?
@choldgraf @chrisjsewell how open are both of you to include new dependencies such as Bokeh?

**Reply by bryanwweber** [reply]
Thanks @damianavila! The Bokeh dependency is only needed if the user calls `glue_bokeh()`, at which point they probably have Bokeh installed. It's imported in the function specifically so that there's not a package-wide dependency on Bokeh 😊

I am working on some more updates as well, to hopefully get rid of the separate API for gluing and pasting Bokeh plots. I'll update later today 😁

**Reply by damianavila** [reply]
> But just from a quick overview

This is why you need to do more than a quick overview... and check the whole thing instead of commenting just looking at the import lines 🤦 

In any case, I am interested to know a little bit more about the policy to accept 3rd party libraries into this project, so my wrong assumptions were actually doing something useful at the very end 😜

**Reply by damianavila** [reply]
> The Bokeh dependency is only needed if the user calls glue_bokeh(), at which point they probably have Bokeh installed.

Btw, I still think you need to catch the low probability event somehow and fail graciously... I can foresee some user trying to use the glue_bokeh without Bokeh installed in their environment (maybe you are deploying this thing somewhere and you forgot to add the Bokeh dependency to your environment) and if you can catch that failure and show an informative message to the user, that should be better feedback than the plain import error, IMHO. Just a thought :wink:

**Reply by choldgraf** [reply]
I think this is really interesting - I doubt that we want Bokeh to be a dependency of MyST-NB (otherwise we'd need to develop a process for deciding when a "commonly used" library should be included or not), but perhaps we can define an extension point like this somehow?

Perhaps we could open an issue around upstreaming the `glue_bokeh` function somehow? I don't want to lose that thread in this discussion! I also know that @chrisjsewell is planning on refactoring some of MyST-NB, including glue functionality, so perhaps this could be a design consideration tehre

**Reply by bryanwweber** [reply]
> Btw, I still think you need to catch the low probability event somehow and fail graciously

I go back and forth about this, I'd be interested in your opinion, or perhaps a more helpful message to the user. Something like:

```python
try:
    import bokeh
except ImportError:
    raise ModuleNotFoundError("Please install bokeh from pip or conda") from None
```

is slightly more pleasant than than the default `ModuleNotFoundError: No module named 'bokeh'`, but it doesn't seem _that_ much more helpful and makes the code more complex... Proceeding to the logical extreme, should we do this for all our dependencies and add dozens of lines of code? If so, shouldn't that be handled with a nicer message in Python itself? Like I said, I go back and forth about which to prefer, I can see both sides 🤷 😄

**Reply by bryanwweber** [reply]
> Perhaps we could open an issue around upstreaming the `glue_bokeh` function somehow? I don't want to lose that thread in this discussion!

Thanks @choldgraf! I'm still working on the code, but I'm happy to either raise this as an issue or a real PR over at MyST-NB for further feedback.

**Reply by bryanwweber** [reply]
I pushed an update to bryanwweber/MyST-NB#1 that moves all the glueing code into the single `glue()` function.

**Reply by damianavila** [reply]
> Like I said, I go back and forth about which to prefer, I can see both sides 🤷 😄

Something interesting from Bokeh itself to manage optional and required dependencies: https://github.com/bokeh/bokeh/blob/main/bokeh/util/dependencies.py

Just adding one more point to the discussion :wink: .

**Reply by choldgraf** [reply]
Interesting, thanks for sharing @damianavila . I'm curious why there's the import for required modules. Are there many users that do not install bokeh in a standard way?

**Reply by damianavila** [reply]
There are multiple Bokeh examples using some 3rd party libraries and those are required by the examples to run but not required for Bokeh to run: https://github.com/bokeh/bokeh/blob/main/bokeh/sampledata/airports.py#L59
Additionally: https://github.com/bokeh/bokeh/blob/main/bokeh/application/handlers/notebook.py#L69, in this case, you only want to install Jupyter components if you really want to use Bokeh code playing with those components :wink:

**Comment by bryanwweber** [comment]
Another approach here: https://github.com/bryanwweber/MyST-NB/pull/2

In this one, I glued the JSON output from Bokeh and inserted HTML/Javascript nodes by using a specific mimetype for this data.

**Reply by bryanwweber** [reply]
The advantage of this approach is that it works transparently with figures with no other changes :tada:

**Reply by bryanwweber** [reply]
I'm happy with the approach here, so I made a PR to the upstream: https://github.com/executablebooks/MyST-NB/pull/335
